### PR TITLE
On-boarding tutorial

### LINKED ISSUE #216: On-boarding tutorial
Related discussion: https://g0v.hackmd.io/eitM7s0bSSeS3kg-MAcpDw#LINE-bot-onboarding-UX

## Target
1. Provide basic usage instruction
2. Experience lookup interaction
3. Setup LIFF permissions

## Trigger
- Join event
- Rich menu

## Interactions

Flex message carousel:
1. When receiving a message from elsewhere / 在他處收到訊息 
2. Long press and share / 長按後分享
3. Select Cofacts to share / 指名 Cofacts
4. Cofacts replies with a crowd-sourced fact-check or chatbot replies / Cofacts 回傳網撰寫的的查證或回應

> Wanna try it out? Just forward a message to me!
> 是不是躍躍欲試呢？試試把訊息傳進來吧！
>
> [按鈕] Simulates forwarding a message / 模擬傳訊息進 Cofacts
 
### 2. Show mock reply

(Reply content without voting)

> [Quick reply] 💡 Cool, I got it! 酷，我懂了！
> [Quick reply]  🤔 What happened? 發生什麼事？

### 3. Provide setup

"Cool, I got it!" -->
> You are smart 😊  你很聰明 😊

"What happened?" -->
> ⬆️ What you see above is a simulation of what you may see after you forward a message to Cofacts. 上面在模擬你傳訊息給我時，會得到的回應。
> 📚 Cofacts has a database of hoax messages and replies. Cofacts 有一個謠言與回應的資料庫。
> 📲 When you send a message to me, I look up the message in our database and return the results I found. 當你傳訊息給我的時候，我會搜尋資料庫然後回傳找到的結果。
> 🆕 If I can't find anything, I will ask you about sending your message to that database. 如果我找不到的話，會徵求你的同意，看看要不要把這個訊息送進資料庫唷。

> To wrap up, please finish your permission settings so that I can provide a smoother experience.
> [按鈕] Setup permission / 權限設定
> [Quick reply]👌 Done! / 我設好了！
> [Quick reply] ⏱️ Later / 之後再設吧
> [Quick reply] ❓ Why setup permission? / 為什麼要設這個？

### 3.5 Explaination
“❓ Why setup permission?” -->
> When I provide hoax-busting replies to you, I would like to ask you for any feedback on the crowd-sourced reply. 
> In order to achieve that, I need to ask for your permission to "send your message to our chatroom".
> The permission will be used to send only this one message of yours back to this particular chatroom.
> You can still use Cofacts without granting me this permission. When we ask for feedback from you, the permission dialog will pop-up again.

### 4. Thank-you

"⏱️ Later" -->
> OK. When we ask for feedback from you, the permission dialog will pop-up again.

> This is the end of the tutorial. Next time when you receive a suspicious message, don't hesitate to forward it to me! 🤗

### PR DESCRIPTION
**Pull request by nonumpa**
Fixes #216

Text setup in rich menu to trigger tutorial should be:
📖 tutorial (en)
📖 教學 (zh-tw)

### CODE REVIEW

File: src/webhook/handlers/tutorial.js [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+ * @returns {object} Flex message object
+ */
+export function createTutorialMessage(data) {
+  const textStep1 = `1. ${t`When receiving a message from elsewhere`}`;
+  const textStep2 = `2. ${t`Long press and share`}`;
+  const textStep3 = `3. ${t`Select Cofacts to share`}`;
+  const textStep4 = `4. ${t`Cofacts replies with a crowd-sourced fact-check or chatbot replies`}`;
+
+  const imageUrlStep1 =
+    'https://uc21ad9bcd486191781c986dfe14.previews.dropboxusercontent.com/p/thumb/AA9R7LsrgHTMDIkz74TvAzOgT5gEcNeCPHWJ_OpIZOYJSBCqzCPMOliCKMe7IRVY_Us_ahRTRgC_d10NkokagS4yRtGo3cGHQfxtEPxRz0Fvo7pxLbEQW3BoM7reJTH0DTbrdku8uTrV3FYz7Evofe55rdRnKGpMPscd9A6YzFeSNvTV31ZwjRvw51utfQG9cwazNCENDVFLeQGl7NANpre3LTKKykTrOuY5t2-VLwYgq6e7fXQGYCGKNuBXWnEljJpPRcD--uDHjsbVtHQFEKt8ap3-lsjoy8AfpmMW619o-pO90Qj3HuRfV8M-_z-JbQS-8cN8Fe0zK-gxMHhRGY_roWKG85vHBePxd7WJDwa--Q/p.png?size=800x600&size_mode=3';
```
Comment (MrOrz) [inline src/webhook/handlers/tutorial.js:112]: We may need a better place for these images 
image

File: src/webhook/handlers/utils.js [hunk 2]
```diff
@@ -563,4 +563,40 @@ function createHighlightContent(text) {
   return result;
 }
 
+/**
+ * @param {*} GetReply
+ * @param {*} GetArticle
```
Comment (MrOrz) [inline src/webhook/handlers/utils.js:568]: Nit: suggest using `reply` and `article` as argument names, and denote that their type is `Reply` and `Article` in rumors-api GraphQL, respectively.

File: src/webhook/handlers/tutorial.js [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+}
+
+/**
+ * @param {object} data
+ * @param {string} message
+ * @returns {object} Flex message object
+ */
+function createPermissionSetupDialog(data, message) {
+  const buttonLabel = t`Setup permission`;
+  const buttonUri = `${process.env.LIFF_URL}/liff/index.html?p=permission`;
```
Comment (MrOrz) [inline src/webhook/handlers/tutorial.js:243]: Hmm by permission settings I actually meant our existing `settings` page. That's why I also put welcome messages in settings LIFF.

File: src/webhook/handlers/tutorial.js [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+ * Fixed inputs that indicate which reply should `tutorial` function return
+ */
+// From rich menu
+export const RICH_MENU_TRIGGER = `📖 ${t`tutorial`}`;
+// From flex message button
+const SIMULATE_FORWARDING_MESSAGE = t`Simulates forwarding a message`;
+// From quick reply, Note: it should be less than 20 charactors
+const PROVIDE_PERMISSION_SETUP = `💡 ${t`Cool, I got it!`}`;
+const EXPLAN_CHATBOT_FLOW_AND_PROVIDE_PERMISSION_SETUP = `🤔 ${t`What happened?`}`;
+const PROVIDE_PERMISSION_SETUP_WITH_EXPLANATION = `❓ ${t`Why permission`}`;
```
Comment (MrOrz) [inline src/webhook/handlers/tutorial.js:19]: Since `Why setup permissions` is more than 20 characters, I suggest we just use `Why`

**Review by MrOrz — COMMENTED** [review]
I have not finish reviewing `handlers/tutorial.js`. Here are some early feedback.

**Review by MrOrz — COMMENTED** [review]
The unit tests are extremely helpful when I read through the code. Thanks for providing the thorough tests cases!

I have some comments regarding wording and how `replies` array can be constructed.

File: src/index.js [hunk 5]
```diff
@@ -45,13 +45,26 @@ if (process.env.NODE_ENV === 'production') {
...
+    })
+  );
 } else {
   app.use(
     require('koa-proxies')('/liff', {
       target: `http://localhost:${process.env.LIFF_DEV_PORT}`,
       logs: true,
     })
   );
+  app.use(
```
[the remark this answers is not in the document]
Comment (nonumpa) [reply src/index.js]: I don't know liff uses proxy to a webpack server for development at that time, just copy and paste the `/liff` part to apply `/static`.

File: src/webhook/handlers/__tests__/__snapshots__/tutorial.test.js.snap [hunk 6]
```diff
@@ -476,8 +476,8 @@ Object {
   "replies": Array [
     Object {
       "altText": "When I provide hoax-busting replies to you, I would like to ask you for any feedback on the crowd-sourced reply.
-In order to achieve that, I need to ask for your permission to \\"send your message to our chatroom\\".
-The permission will be used to send only this one message of yours back to this parti⋯⋯
+In order to achieve that, I need to ask for your permission to \\"Send messages to chats\\".
+The permission will be used to send only this one message of yours back to this particular chatr⋯⋯
```
Comment (nonumpa) [inline src/webhook/handlers/__tests__/__snapshots__/tutorial.test.js.snap:480]: Text will ellipsis before assigning to altText, so the snapshot looks a little weird.

**Review by MrOrz — APPROVED** [review]
LGTM! Let's merge!

**Pull request merged**